adapter/stdio/stdio.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stdio // import "istio.io/mixer/adapter/stdio"
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	multierror "github.com/hashicorp/go-multierror"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"istio.io/mixer/adapter/stdio/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/template/logentry"
+	"istio.io/mixer/template/metric"
+)
+
+type (
+	zapBuilderFn func(outputPath string) (*zap.Logger, error)
+
+	builder struct {
+		zapBuilder zapBuilderFn // indirection to allow override in tests
+	}
+
+	handler struct {
+		zapper         *zap.Logger
+		severityLevels map[string]zapcore.Level
+		metricLevel    zapcore.Level
+		omitEmpty      bool
+	}
+)
+
+// ensure our types implement the requisite interfaces
+var _ logentry.HandlerBuilder = builder{}
+var _ logentry.Handler = &handler{}
+var _ metric.HandlerBuilder = builder{}
+var _ metric.Handler = &handler{}
+
+///////////////// Configuration Methods ///////////////
+
+func (b builder) Build(cfg adapter.Config, _ adapter.Env) (adapter.Handler, error) {
+	c := cfg.(*config.Params)
+
+	outputPath := "stdout"
+	if c.LogStream == config.STDERR {
+		outputPath = "stderr"
+	}
+
+	zapLogger, err := b.zapBuilder(outputPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not build logger: %v", err)
+	}
+
+	sl := make(map[string]zapcore.Level)
+	for k, v := range c.SeverityLevels {
+		sl[k] = mapConfigLevel(v)
+	}
+
+	return &handler{
+		omitEmpty:      c.OmitEmptyFields,
+		severityLevels: sl,
+		metricLevel:    mapConfigLevel(c.MetricLevel),
+		zapper:         zapLogger,
+	}, nil
+}
+
+func mapConfigLevel(l config.Params_Level) zapcore.Level {
+	if l == config.WARNING {
+		return zapcore.WarnLevel
+	} else if l == config.ERROR {
+		return zapcore.ErrorLevel
+	}
+	return zapcore.InfoLevel
+}
+
+func (builder) ConfigureLogEntryHandler(map[string]*logentry.Type) error {
```

Comment:
Hmm.. I think that is too risky since it is prone to spelling mistake errors. Operator might miss spell the method and we will silently stop serving them type information. Too risky I think.
IMO, being strict is better here, writing an empty method is no big deal + adapter developer and his code reviewers will exactly know the reason for the method to be empty.

As I am writing the developer guide, I realize we are doing enough black magic and adding more of a black boxed concept will only add to confusion.

Having said that, if we really want this optional implementation of HandlerBuilder and still have a strict validation, we should ask the adapter developer to explicitly state in BuilderInfo (aka adapter.Info) that he wants just 'Handler' interface part of a particular mesh function. So, declaring just the mesh function name in the BuilderInfo will not be sufficient, they will have to give something in the lines of:
"metric.TemplateName,Handler" or "metric.TemplateName,Handler, HandlerBuilder". Which I think is also a complex concept to understand.